src/utils/include/utils/tensor.hpp

```diff
@@ -0,0 +1,194 @@
+    assert(valid_indices(indices));
+    return m_data[calc_linear_index(indices)];
+  }
+
+  /**
+   * @brief Bounds checked element access.
+   * @param indices index for each dimension
+   */
+  template <typename... size_ts> const_reference at(size_ts... indices) const {
+    return at({static_cast<std::size_t>(indices)...});
+  }
+
+  template <typename... size_ts> reference at(size_ts... indices) {
+    return at({static_cast<std::size_t>(indices)...});
+  }
+
+  template <typename Container>
+  const_reference at(Container const &indices) const {
+    validate_indices(indices);
+    return operator()(indices);
+  }
+
+  template <typename Container> reference at(Container const &indices) {
+    validate_indices(indices);
+    return operator()(indices);
+  }
+
+  const_reference at(std::initializer_list<std::size_t> const &indices) const {
+    validate_indices(indices);
+    return operator()(indices);
+  }
+
+  reference at(std::initializer_list<std::size_t> const &indices) {
+    validate_indices(indices);
+    return operator()(indices);
+  }
+
+  reference front() { return *begin(); }
+
+  const_reference front() const { return *cbegin(); }
+
+  reference back() { return m_data.back(); }
+
+  const_reference back() const { return m_data.back(); }
+
+  iterator begin() noexcept { return &(*m_data.begin()); };
+
+  const_iterator begin() const noexcept { return &(*m_data.begin()); };
+
+  const_iterator cbegin() const noexcept { return &(*m_data.begin()); };
+
+  iterator end() noexcept { return &(*m_data.end()); };
+
+  const_iterator end() const noexcept { return &(*m_data.end()); };
+
+  const_iterator cend() const noexcept { return &(*m_data.end()); };
+
+  pointer data() noexcept { return m_data.data(); }
+
+  const_pointer data() const noexcept { return m_data.data(); }
+
+  auto rank() const noexcept { return m_extents.size(); }
+
+  auto const &extents() const noexcept { return m_extents; }
+
+private:
+  /**
+   * @brief Check that indices are valid.
+   * The number of given indices as well as being within the extents is checked.
+   * @param indices Indices to be checked.
+   * @returns boolean true if valid, false else.
+   * @tparam Type of container holding the indices.
+   */
+  template <typename Container>
+  bool valid_indices(Container const &indices) const {
+    if (indices.size() != m_extents.size()) {
+      return false;
+    }
+    // check if any index exceeds an extend.
+    auto const res =
+        boost::mismatch(indices, m_extents, std::less<std::size_t>());
+    return res.first == indices.end();
+  }
+
+  /**
+   * @brief Throw if invalid indices.
+   * @param indices Indices to be checked.
+   * @tparam Type of container holding the indices.
+   * @throws std::runtime_error if indices are invalid.
+   */
+  template <typename Container>
+  void validate_indices(Container const &indices) const {
+    if (not valid_indices(indices)) {
+      throw std::runtime_error("Invalid indices");
```

Comment:
`std::vector` throws `std::out_of_range`

Reply:
ok, but it's not only checking for out of range

Reply:
Because it also checks the length? It's still what I would expect